### PR TITLE
consolidate datafusion docs with sphinx

### PR DESCRIPTION
**Pull request by houqp**
Co-authored-by: Jiayu Liu <Jimexist@users.noreply.github.com>

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/837
Closes https://github.com/apache/arrow-datafusion/issues/18
Relates to https://github.com/apache/arrow-datafusion/issues/18

 # Rationale for this change

This PR builds on top of [the great work](https://github.com/apache/arrow/pull/10982) @Jimexist did with our Python binding doc.

In summary, the suggestion from the arrow PR was to host the Python binding doc in our a website dedicated to datafusion instead of in Arrow's documentation page. Then link to the Datafusion documentation from Arrow home page's top banner.

Following the plan outlined at https://github.com/apache/arrow/pull/10982#issuecomment-911233468, this PR consolidates all docs into a single doc website so we can use it as the home page for Datafusion (to be hosted at https://arrow.apache.org/datafusion). The Python binding doc prepared by @Jimexist will be hosted as one of the sections in the Datafusion doc.

Once this PR is merged, I will generate the docs and send the full html build as a PR to the arrow-site repo.

# What changes are included in this PR?

* added python doc
* combined, cli, user-guide and specification docs into a single datafusion doc

# Are there any user-facing changes?

No

Here is what the documentation looks like:

df-doc

df-doc-2

### CODE REVIEW

**Comment by houqp** [comment]
I believe this is the last missing piece before we can prepare for the 5.1.0 release.

File: docs/Makefile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (houqp) [inline docs/Makefile:1]: generated by sphinx

File: docs/make.bat [hunk 2]
```diff
@@ -0,0 +1,52 @@
+@rem Licensed to the Apache Software Foundation (ASF) under one
```
Comment (houqp) [inline docs/make.bat:1]: generated by sphinx

File: docs/source/_static/theme_overrides.css [hunk 3]
```diff
@@ -0,0 +1,93 @@
+/**
```
Comment (houqp) [inline docs/source/_static/theme_overrides.css:1]: copied from arrow

File: docs/source/_templates/docs-sidebar.html [hunk 4]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (houqp) [inline docs/source/_templates/docs-sidebar.html:1]: copied from arrow

File: docs/README.md [hunk 5]
```diff
@@ -17,15 +17,19 @@
...
+## Dependencies
 
-## Generate HTML
+It's recommended to install build dependencies and build the the documentation
+inside a Python virtualenv.
 
-To generate the user guide in HTML format, run the following commands:
+- Python
+- `pip install -r requirements.txt`
+- Datafusion python package. You can install the latest version by running `maturin develop` inside `../python` directory.
```
Comment (alamb) [inline docs/README.md:29]: I tried these instructions locally and one other thing I had to do was to run `pip install maturin` in my venv to be able to run `maturn develop`  

So perhaps we should also suggest adding `pip install maturin` alongside `maturin develop`

**Review by alamb — APPROVED** [review]
I took it for a spin and it worked great 👍  Really nice @houqp 

Screen Shot 2021-09-13 at 1 37 50 PM

File: docs/README.md (hunk 5 above)
Comment (houqp) [reply docs/README.md:29]: added maturin to requirements.txt

**Comment by houqp** [comment]
@Jimexist since you put in a lot of efforts into building out the python doc, do you mind taking a look at this PR?

File: docs/README.md [hunk 6]
```diff
@@ -17,15 +17,19 @@
...
 
-```bash
-cargo install mdbook
-mdbook build
+## Build
+
+```
```
Comment (Jimexist) [inline docs/README.md:33]:
```suggestion
```bash
```

**Review by Jimexist — APPROVED** [review]

**Comment by Jimexist** [comment]
generally LGTM, i might want to improve on the `requirements.txt` for reproducible builds but that's to be done in subsequent PRs.

**Comment by Jimexist** [comment]
i think that https://github.com/apache/arrow-datafusion/pull/977 can go after you merge this - which i'll try to update `.ast` files instead

**Review by Dandandan — APPROVED** [review]

**Comment by alamb** [comment]
🎉 📖

**Pull request merged**